consensus/core/src/linearizer.rs

```diff
@@ -138,6 +144,55 @@ impl Linearizer {
         (sub_dag, commit)
     }
 
+    /// Calculates the commit's timestamp. If the median based timestamp calculation is enabled, then the timestamp will be calculated as the median of
+    /// leader's committed strong ancestors (leader.round - 1). Otherwise, the leader's timestamp will be used. To ensure that commit timestamp monotonicity is
+    /// respected it is compared against the `last_commit_timestamp_ms` and the maximum of the two is returned.
+    pub(crate) fn calculate_commit_timestamp(
+        context: &Context,
+        leader_block: &VerifiedBlock,
+        last_commit_timestamp_ms: BlockTimestampMs,
+        to_commit: &[VerifiedBlock],
+    ) -> BlockTimestampMs {
+        let timestamp_ms = if context
+            .protocol_config
+            .consensus_median_based_commit_timestamp()
+        {
+            // Attention should be paid that we are calculating the timestamp only using the leader's committed strong ancestors which should be all the blocks
+            // or round leader.round - 1. Not necessary to search and recheck here the leader's ancestors as by the protocol rules it's impossible to commit any block
+            // of round leader.round - 1 unless it is a leader's strong ancestor.
+            let timestamps = to_commit
+                .iter()
+                .filter_map(|block| {
+                    if block.round() == leader_block.round() - 1 {
+                        Some(block.timestamp_ms())
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            // The only reason for the median to be `None` would be when there is only one authority in commitee and the only block that gets committed is the leader's.
```

Comment:
Another alternative is to get ts from ancestors of the leader regard of if they are in the commit, then take the median. This way it might be more consistent.